plugins/AudioFileProcessor/AudioFileProcessor.cpp

```diff
@@ -292,15 +292,24 @@ QString AudioFileProcessor::nodeName() const
 
 
 
-int AudioFileProcessor::getBeatLen( NotePlayHandle * _n ) const
+auto AudioFileProcessor::beatLen(NotePlayHandle* note) const -> int
 {
+	// If we can play indefinitely, use the default beat note duration
+	if (static_cast<SampleBuffer::LoopMode>(m_loopModel.value()) != SampleBuffer::LoopMode::Off) { return 0; }
+
+	// Otherwise, use the remaining sample duration
 	const auto baseFreq = instrumentTrack()->baseFreq();
-	const float freq_factor = baseFreq / _n->frequency() *
-			Engine::audioEngine()->processingSampleRate() / Engine::audioEngine()->baseSampleRate();
+	const auto freqFactor = baseFreq / note->frequency()
+		* Engine::audioEngine()->processingSampleRate()
+		/ Engine::audioEngine()->baseSampleRate();
 
-	return static_cast<int>( floorf( ( m_sampleBuffer.endFrame() - m_sampleBuffer.startFrame() ) * freq_factor ) );
-}
+	const auto startFrame = m_nextPlayStartPoint >= m_sampleBuffer.endFrame()
+		? m_sampleBuffer.startFrame()
+		: m_nextPlayStartPoint;
+	const auto duration = m_sampleBuffer.endFrame() - startFrame;
 
+	return static_cast<int>(floorf(duration * freqFactor));
```

Comment:
```suggestion
	return static_cast<int>(std::floorf(duration * freqFactor));
```

Reply:
I was leaning towards leaving existing code alone, but I've basically rewritten the whole method at this point, so I'll go ahead and make the change.